### PR TITLE
added GdbmIterator to iterate over the list of keys in the database

### PR DESCRIPTION
**Pull request by FrancoisGib**
I added an iterator to iterate over the keys of a gdbm database
I also implemented the Drop trait to correctly free the datum when the iterator is freed

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -154,6 +153,49 @@ bitflags! {
...
+    db_handle: GDBM_FILE,
+    current_key: Option<datum>,
+}
+
+impl Iterator for GdbmIterator {
+    type Item = Result<String, GdbmError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let content = if let Some(current_key) = self.current_key {
+            let next_key = unsafe { gdbm_nextkey(self.db_handle, current_key) };
```
Comment (cholcombe973) [inline src/lib.rs:167]: Would you mind documenting the assumptions here for the unsafe blocks?

File: src/lib.rs [hunk 2]
```diff
@@ -319,30 +362,14 @@ impl Gdbm {
...
-    // }
-    //
+    /// Return an iterator over all keys in the database.
+    pub fn iter(&self) -> GdbmIterator {
+        GdbmIterator {
+            db_handle: self.db_handle,
+            current_key: None,
+        }
+    }
+
```
Comment (cholcombe973) [inline src/lib.rs:372]: I think it would be good to introduce a few integration tests that look for edge cases

**Review by cholcombe973 — CHANGES_REQUESTED** [review]
Thanks for doing this! Looks like I had an ancient comment in the code about creating an iterator but never got back to it. The unsafe blocks look correct to me. I think it would be worth it to write a few small integration tests both for documentation purposes to show usage but also to run valgrind against and just double check the unsafe blocks.

**Comment by akiradeveloper** [comment]
@FrancoisGib Do you have time to finish this? If not, I will take over the remaining tasks.

**Comment by cholcombe973** [comment]
@akiradeveloper could you test out this branch and let us know what your results are?

**Comment by akiradeveloper** [comment]
@cholcombe973 

I added this test but it failed at the last part after deleting one elem.

Here is the error message. Any idea?

```
[tests/test.rs:62:9] &k = Err(
    Error(
        "invalid utf-8 sequence of 1 bytes from index 3",
    ),
)
```

```rust
#[test]
fn list_test() {
    use std::collections::HashSet;

    let f = tempfile::NamedTempFile::new().unwrap();
    let db = gdbm::Gdbm::new(f.path(),
                                 0,
                                 gdbm::Open::NEWDB,
                                 S_IRUSR | S_IWUSR)
        .expect("Gdbm::new");

    // Empty
    let mut h = HashSet::new();
    for k in db.iter() {
        h.insert(k.unwrap());
    }
    assert_eq!(h.len(), 0);

    // Add 3 elems
    db.store("c", "ccc", false).unwrap();
    db.store("b", "bbb", false).unwrap();
    db.store("a", "aaa", false).unwrap();
    let mut h = HashSet::new();
    for k in db.iter() {
        h.insert(k.unwrap());
    }
    assert_eq!(h, maplit::hashset!{"a".to_string(), "b".to_string(), "c".to_string()});

    // Remove one elem
    db.delete("b").unwrap();
    let mut h = HashSet::new();
    for k in db.iter() {
        dbg!(&k);
        h.insert(k.unwrap());
    }
    assert_eq!(h, maplit::hashset!{"a".to_string(), "c".to_string()});

}
```
